### PR TITLE
NAS-127350 / 24.10 / Add ability to filter by keys in list objects

### PR DESCRIPTION
**Pull request by anodos325**
There are some cases where API user may wish to query an endpoint based on nested values within a list. For example

[["local_groups.*.name", "=", "bad"]] passed to privilege.query to return all privileges where the group "bad" is present in the list of local groups to which the privilege applies.

### CODE REVIEW

File: src/middlewared/middlewared/apidocs/templates/websocket/query.md [hunk 1]
```diff
@@ -132,6 +132,28 @@ Javascript:
...
+    ]
+
+Alternatively, the index may be omitted, which will result in the result returning if any array member has a key matching the
+value.
+
+Javascript:
+
+    :::javascript
+    [
+      ["local_groups.name", "=", "myuser"],
```
Comment (themylogin) [inline src/middlewared/middlewared/apidocs/templates/websocket/query.md:153]: Can we be writing this in a less ambiguous way? Like `local_groups.*.name`

Comment (anodos325) [reply src/middlewared/middlewared/apidocs/templates/websocket/query.md]: Yes. That's probably a better idea.

**Review by themylogin — APPROVED** [review]
Thank you!

File: src/middlewared/middlewared/utils/__init__.py [hunk 2]
```diff
@@ -96,9 +100,40 @@ def partition(s):
...
+        elif isinstance(cur, (list, tuple)):
+            if not left.isdigit():
+                # return all members and the remaining portion of path
+                if left == '*':
+                    return FilterGetResult(result=cur, key=right, done=False)
+
+                raise ValueError(f'{left}: must be array index or wildcard character')
+
+            left = int(left)
+            cur = cur[left] if left < len(cur) else None
```
Comment (sonicaj) [inline src/middlewared/middlewared/utils/__init__.py:119]: if `cur` is `None`, i don't think there is any point in continuing the loop ?

**Review by sonicaj — APPROVED** [review]

**Pull request merged**